### PR TITLE
[DCP] Support Decode Context Parallel (DCP) for GQA with FlashAttention

### PR DESCRIPTION
**Pull request by FENP**
## Purpose
This PR adds Decode Context Parallel (DCP) support for GQA following PR https://github.com/vllm-project/vllm/pull/23734. Current implementation based on FlashAttention. 

Unlike MLA inference,  GQA (with FlashAttention) does not distinguish between prefill and decode during forward pass. To support DCP, this PR separately computes the attention scores for the context and query KV within a sequence and then merges the results.
```text  
  # |- tokenA -|......................|-- newTokens ---|
  # |---------- context_len ----------|-- query_len ---|
```
- For the query, no collective communication is required among the DCP group. 
- For context, the KV is distributed across different DCP ranks. This PR follows the DCP decode approach from MLA, i.e., all-gathering Q and lse, then correcting the attn out before performing reduce-scatter.

## Test Plan
Qwen/Qwen3-235B-A22B-Instruct-2507
```shell
vllm serve Qwen/Qwen3-235B-A22B-Instruct-2507 --gpu-memory-utilization 0.9 --tensor-parallel-size 8 --decode-context-parallel-size 2
```

## Test Result
- **KV Cache Size**
```shell
[kv_cache_utils.py:859] Multiplying the GPU KV cache size by the dcp_world_size 2.
[kv_cache_utils.py:864] GPU KV cache size: 1,723,584 tokens
[kv_cache_utils.py:868] Maximum concurrency for 262,144 tokens per request: 6.57x
```
- **gsm8k eval**
```text
TP8
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8836|±  |0.0062|
|     |       |strict-match    |     5|exact_match|↑  |0.8628|±  |0.0067|

TP8DCP2
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8988|±  |0.0059|
|     |       |strict-match    |     5|exact_match|↑  |0.8870|±  |0.0062|
```

Additionally, I think we can reduce latency by overlapping query computation with context communication.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/attention/backends/flash_attn.py [hunk 1]
```diff
@@ -288,8 +303,32 @@ def schedule(batch_size, cu_query_lens, max_query_len, seqlens,
...
+        prefix_scheduler_metadata = None
+
+        if self.dcp_world_size > 1:
+            query_kv_lens_cpu = common_attn_metadata.query_start_loc_cpu[1:] \
+                - common_attn_metadata.query_start_loc_cpu[:-1]
+            dcp_context_kv_lens_cpu = seq_lens_cpu - query_kv_lens_cpu
+            dcp_context_kv_lens_cpu = dcp_context_kv_lens_cpu \
+                // self.dcp_world_size + (self.dcp_rank \
+                <= (dcp_context_kv_lens_cpu-1) % self.dcp_world_size)
+            dcp_context_kv_lens = dcp_context_kv_lens_cpu.to(self.device)
```
Comment (youzhedian) [inline vllm/v1/attention/backends/flash_attn.py:356]: nit: `non_blocking=True` is a little better?

File: vllm/v1/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -367,6 +404,8 @@ def use_cascade_attention(self, *args, **kwargs) -> bool:
 
 class FlashAttentionImpl(AttentionImpl):
 
+    can_return_lse_for_decode: bool = True
```
Comment (youzhedian) [inline vllm/v1/attention/backends/flash_attn.py:407]: unused?

**Review by youzhedian — CHANGES_REQUESTED** [review]
plz add GQA+DCP ut in `test_context_parallel.py`.

Others, LGTM. Clean and good job!

File: vllm/v1/attention/backends/flash_attn.py (hunk 1 above)
Comment (youzhedian) [reply vllm/v1/attention/backends/flash_attn.py:356]: Qwen2.5-3B: /mnt/moonfs/public-models-ksyun/Qwen/Qwen2.5-3B is GQA2 & small, is suit for ut

File: vllm/v1/attention/backends/flash_attn.py [hunk 3]
```diff
@@ -581,6 +635,87 @@ def forward(
...
+                v_descale=v_descale,
+            )
+        # FA returns LSE in shape [ H, B ] but cp_lse_ag_out_rs wants [ B, H ]
+        context_lse = context_lse.transpose(0, 1)
+        context_attn_out, context_lse = cp_lse_ag_out_rs(
+                                            context_attn_out, 
+                                            context_lse, 
+                                            get_dcp_group(),
+                                            return_lse=True)
+        context_lse = context_lse.transpose(0, 1).contiguous()
```
Comment (youkaichao) [inline vllm/v1/attention/backends/flash_attn.py]: variable naming in this part is confusing, `context_lse` is shadowed multiple times.

File: vllm/config/__init__.py [hunk 4]
```diff
@@ -1367,6 +1367,17 @@ def verify_with_parallel_config(
 
             if self.use_async_output_proc:
                 self.use_async_output_proc = False
+        
+        decode_context_parallel_size = \
+            parallel_config.decode_context_parallel_size
+        if decode_context_parallel_size > 1 and not self.use_mla:
+            total_num_kv_heads = self.get_total_num_kv_heads()
+            max_dcp_size = tensor_parallel_size // total_num_kv_heads
```
Comment (youkaichao) [inline vllm/config/__init__.py]: assert `tensor_parallel_size > total_num_kv_heads`, otherwise we don't need dcp.

**Review by youkaichao — COMMENTED** [review]
please fix pre-commit linter error, and add tests. otherwise LGTM, thanks for the great work!

cc @LucasWilkinson to give the final sign off.

File: tests/models/registry.py [hunk 5]
```diff
@@ -313,7 +313,10 @@ def check_available_online(
                                        transformers_version_reason="HF model uses remote code that is not compatible with latest Transformers",  # noqa: E501
                                        trust_remote_code=True),
     "Qwen2ForCausalLM": _HfExamplesInfo("Qwen/Qwen2-0.5B-Instruct",
-                                        extras={"2.5": "Qwen/Qwen2.5-0.5B-Instruct"}), # noqa: E501
+                                        extras={
+                                            "2.5": "Qwen/Qwen2.5-0.5B-Instruct",
+                                            "2.5-3b": "Qwen/Qwen2.5-3B-Instruct"
+                                        }),
```
Comment (youkaichao) [inline tests/models/registry.py]: why do we need to change this file?

File: tests/distributed/test_context_parallel.py [hunk 6]
```diff
@@ -222,14 +222,15 @@ def _compare_cp_with_tp(
 
 
 CP_TEXT_GENERATION_MODELS = {
-    # [MLA attention only]
     "deepseek-ai/DeepSeek-V2-Lite-Chat": CPTestSettings.detailed(),
+    "Qwen/Qwen2.5-3B-Instruct": CPTestSettings.detailed(tp_base=8),
```
Comment (youkaichao) [inline tests/distributed/test_context_parallel.py]: `tp_base=8` will require 8 GPUs, we don't have 8 GPUs in ci.

File: tests/models/registry.py (hunk 5 above)
Comment (FENP) [reply tests/models/registry.py]: > why do we need to change this file?

I add Qwen/Qwen2.5-3B-Instruct in tests/distributed/test_context_parallel.py to test DCP for GQA. It needs find model info from this file.
https://github.com/vllm-project/vllm/blob/e017120ed146cc3069d18428322d67881cb95e67/tests/distributed/test_context_parallel.py#L123-L124

File: tests/distributed/test_context_parallel.py (hunk 6 above)
Comment (FENP) [reply tests/distributed/test_context_parallel.py]: > `tp_base=8` will require 8 GPUs, we don't have 8 GPUs in ci.

Fix.

File: vllm/v1/attention/backends/flash_attn.py (hunk 1 above)
Comment (FENP) [reply vllm/v1/attention/backends/flash_attn.py:356]: > nit: `non_blocking=True` is a little better?

Now `dcp_context_kv_lens_cpu` is not a pinned memory tensor :). Maybe we could reuse seq_lens_cpu?
I'm not sure if this will cause confusion in variable naming.

File: vllm/v1/attention/backends/flash_attn.py [hunk 7]
```diff
@@ -287,8 +302,31 @@ def schedule(batch_size, cu_query_lens, max_query_len, seqlens,
...
+                - common_attn_metadata.query_start_loc_cpu[:-1]
+            dcp_context_kv_lens_cpu = seq_lens_cpu - query_kv_lens_cpu
+            dcp_context_kv_lens_cpu = dcp_context_kv_lens_cpu \
+                // self.dcp_world_size + (self.dcp_rank \
+                <= (dcp_context_kv_lens_cpu-1) % self.dcp_world_size)
+            dcp_context_kv_lens = dcp_context_kv_lens_cpu.to(self.device)
+            max_dcp_context_kv_len = dcp_context_kv_lens.max().item()
 
-        if use_cascade:
+            scheduler_metadata = schedule(batch_size=num_reqs,
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/flash_attn.py]: I think the head count for the `schedule` call will be wrong in the dcp case since right now its:

```
                return get_scheduler_metadata(
                    ...
                    num_heads_q=self.num_heads_q,
                    num_heads_kv=self.num_heads_kv,
                    ...
                )
```

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for the contribution! This is awesome! Left one comment.

Side note: I would like us to start thinking about if there is a backend agnostic way of doing this though since FA is basically only Ampere and Hopper at this point (FlashInfer is the premier kernel for blackwell);

Ill think about about it more; but my initial though would be we can just wrap the backends similar to what we do for [ChunkedLocalAttention](https://github.com/vllm-project/vllm/blob/4e5affeaa14220b5699c667e6c3ee5402b4284ef/vllm/attention/layers/chunked_local_attention.py#L71-L76) and [YOCO fast prefill backend](https://github.com/vllm-project/vllm/blob/4e5affeaa14220b5699c667e6c3ee5402b4284ef/vllm/v1/worker/gpu_model_runner.py#L3203-L3206)

This would mean having to add support for the backends returning LSE. This may be easier if we pull `reshape_and_cache_flash` and `concat_and_cache_mla` out of the backend per https://github.com/vllm-project/vllm/issues/24678 / https://github.com/vllm-project/vllm/issues/24620 . Then roughly we might be able to do something like:

```
class DCPAttentionBuilder(underlying_builder):  # type: ignore
    class DCPAttentionMetadata(underlying_metadata):
        context_metadata: AttentionMetadata
        new_tokens_metadata: AttentionMetadata

    def build(self,
                common_prefix_len: int,
                common_attn_metadata: CommonAttentionMetadata,
                fast_build: bool = False) -> AttentionMetadata:
        common_attn_metadata_context = adjust_for_dcp_context(common_attn_metadata)
        common_attn_metadata_context.causal = False

        common_attn_metadata_new_tokens = common_attn_metadata
        common_attn_metadata_new_tokens.seq_lens = query_lens
        
        return DCPAttentionMetadata(
            context_metadata=super().build(common_prefix_len, common_attn_metadata_context, fast_build),
            new_tokens_metadata=super().build(common_prefix_len, common_attn_metadata, fast_build)
        )
        
class DCPImpl(underlying_impl):
    ...
    
    def forward(
        self,
        layer: torch.nn.Module,
        query: torch.Tensor,
        key: torch.Tensor,
        value: torch.Tensor,
        kv_cache: torch.Tensor,
        attn_metadata: DCPAttentionBuilder.DCPAttentionMetadata,
        ...
    ) -> torch.Tensor:
        query_across_dcp = get_dcp_group().all_gather(query, dim=1)
        context_out = super().forward(
            layer,
            query_across_dcp,
            key,
            value,
            kv_cache,
            attn_metadata.context_metadata,
            return_lse=True,
            ...
        )
        
        context_attn_out_cor, context_lse_cor = cp_lse_ag_out_rs(...)

        # Somehow make sure this will use key/value instead of kv_cache
        #  we could detect this based on the fact that kv_len == q_len in this
        #  case so there's really no reason to use the kv_cache
        new_tokens_out = super().forward(
            layer,
            query,
            key,
            value,
            kv_cache,
            attn_metadata.new_tokens_metadata,
            return_lse=True,
            ...
        )

        merge_attn_states(...)

        return merge_attn_states(...)
```


This is very rough; I think its pretty tricky so I think it still makes sense to land this FA only PR version first. I just want to get the conversation started on how we might make this more general before we go start repeatly applying similar changes to all backends.

File: vllm/v1/attention/backends/flash_attn.py (hunk 7 above)
Comment (FENP) [reply vllm/v1/attention/backends/flash_attn.py]: > I think the head count for the `schedule` call will be wrong in the dcp case since right now its:
> 
> ```
>                 return get_scheduler_metadata(
>                     ...
>                     num_heads_q=self.num_heads_q,
>                     num_heads_kv=self.num_heads_kv,
>                     ...
>                 )
> ```

That's right. After `get_dcp_group().all_gather(query, dim=1)`, `num_heads_q` should be `self.num_heads_q * self.dcp_world_size`.  Fix done.

**Comment by FENP** [comment]
> Thanks for the contribution! This is awesome! Left one comment.
> 
> Side note: I would like us to start thinking about if there is a backend agnostic way of doing this though since FA is basically only Ampere and Hopper at this point (FlashInfer is the premier kernel for blackwell);
> 
> Ill think about about it more; but my initial though would be we can just wrap the backends similar to what we do for [ChunkedLocalAttention](https://github.com/vllm-project/vllm/blob/4e5affeaa14220b5699c667e6c3ee5402b4284ef/vllm/attention/layers/chunked_local_attention.py#L71-L76) and [YOCO fast prefill backend](https://github.com/vllm-project/vllm/blob/4e5affeaa14220b5699c667e6c3ee5402b4284ef/vllm/v1/worker/gpu_model_runner.py#L3203-L3206)
> 
> This would mean having to add support for the backends returning LSE. This may be easier if we pull `reshape_and_cache_flash` and `concat_and_cache_mla` out of the backend per #24678 / #24620 . Then roughly we might be able to do something like:
> 
> ```
> class DCPAttentionBuilder(underlying_builder):  # type: ignore
>     class DCPAttentionMetadata(underlying_metadata):
>         context_metadata: AttentionMetadata
>         new_tokens_metadata: AttentionMetadata
> 
>     def build(self,
>                 common_prefix_len: int,
>                 common_attn_metadata: CommonAttentionMetadata,
>                 fast_build: bool = False) -> AttentionMetadata:
>         common_attn_metadata_context = adjust_for_dcp_context(common_attn_metadata)
>         common_attn_metadata_context.causal = False
> 
>         common_attn_metadata_new_tokens = common_attn_metadata
>         common_attn_metadata_new_tokens.seq_lens = query_lens
>         
>         return DCPAttentionMetadata(
>             context_metadata=super().build(common_prefix_len, common_attn_metadata_context, fast_build),
>             new_tokens_metadata=super().build(common_prefix_len, common_attn_metadata, fast_build)
>         )
>         
> class DCPImpl(underlying_impl):
>     ...
>     
>     def forward(
>         self,
>         layer: torch.nn.Module,
>         query: torch.Tensor,
>         key: torch.Tensor,
>         value: torch.Tensor,
>         kv_cache: torch.Tensor,
>         attn_metadata: DCPAttentionBuilder.DCPAttentionMetadata,
>         ...
>     ) -> torch.Tensor:
>         query_across_dcp = get_dcp_group().all_gather(query, dim=1)
>         context_out = super().forward(
>             layer,
>             query_across_dcp,
>             key,
>             value,
>             kv_cache,
>             attn_metadata.context_metadata,
>             return_lse=True,
>             ...
>         )
>         
>         context_attn_out_cor, context_lse_cor = cp_lse_ag_out_rs(...)
> 
>         # Somehow make sure this will use key/value instead of kv_cache
>         #  we could detect this based on the fact that kv_len == q_len in this
>         #  case so there's really no reason to use the kv_cache
>         new_tokens_out = super().forward(
>             layer,
>             query,
>             key,
>             value,
>             kv_cache,
>             attn_metadata.new_tokens_metadata,
>             return_lse=True,
>             ...
>         )
> 
>         merge_attn_states(...)
> 
>         return merge_attn_states(...)
> ```
> 
> This is very rough; I think its pretty tricky so I think it still makes sense to land this FA only PR version first. I just want to get the conversation started on how we might make this more general before we go start repeatly applying similar changes to all backends.

Thank you for commenting! I really agree with your wrap abstract. Do we need the backend to provide two types of forward functions, one using KV Cache and the other using key and value directly? 

Additionally, if we need to support MLA DCP compatibility, we might need to do further abstraction. Maybe we can start with metadata first.

**Comment by FENP** [comment]
Hi @LucasWilkinson . Could you re-review this PR and give the final sign off ? Thanks!

**Review by LucasWilkinson — APPROVED** [review]
Apologies for the delay! LGTM! Thanks for the contribution!
